File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        def merge(l1: List[Optional[ListNode]], l2: List[Optional[ListNode]]) -> Optional[ListNode]:
            tmp = dummy = ListNode()
            while l1 and l2:
                if l1.val < l2.val:
                    dummy.next = l1
                    l1 = l1.next
                else:
                    dummy.next = l2
                    l2 = l2.next
                dummy = dummy.next
            if l1:
                dummy.next = l1
            if l2: 
                dummy.next = l2
            return tmp.next
        
        
        if not len(lists):
            return None
            
        resultList = lists[0]
        for l in range(1, len(lists)):
            currList = lists[l]
            resultList = merge(resultList, currList)
        return resultList

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode res(0);
        ListNode *tmp = &res;

        while (list1 && list2) {
            ListNode ** mark;
            if (list1->val < list2->val) {
                mark = &list1;
            } else {
                mark = &list2;
            }
            tmp->next = *mark;
            *mark = (*mark)->next;
            tmp = tmp->next;
       }
       if (list1) tmp->next = list1;
       if (list2) tmp->next = list2;

       return res.next;
    }
};